### PR TITLE
contrib/labstack/echo.v4: add WithErrorTranslator Option

### PR DESCRIPTION
**Pull request by mattscamp**
### What does this PR do?

Adds a `WithErrorTranslator` Option that allows defining a translator of custom errors to the required format.

### Motivation

We use a custom error type across our company that cannot be translated to an `echo.HTTPError` which results in all our status codes set as 500.
 
### Reviewer's Checklist
- [x] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.

### CODE REVIEW

**Comment by mattscamp** [comment]
Hey @ahmed-mez , if there is anything I can do to expedite this PR, let me know!

**Comment by ahmed-mez** [comment]
Hey @mattscamp - thanks for the contribution! The PR is triaged and the corresponding team will review in the coming days.
cc @rarguelloF

File: contrib/labstack/echo.v4/option.go [hunk 1]
```diff
@@ -33,11 +34,15 @@ type Option func(*config)
...
 
+// CustomErrorFunc attempts to translate a non-standard echo error to the required format
+type CustomErrorFunc func(err error) *echo.HTTPError
+
 func defaults(cfg *config) {
 	cfg.serviceName = namingschema.NewDefaultServiceName(defaultServiceName).GetName()
 	cfg.analyticsRate = math.NaN()
 	cfg.isStatusError = isServerError
 	cfg.headerTags = globalconfig.HeaderTagMap()
+	cfg.customErrorFunc = func(err error) *echo.HTTPError { return nil }
```
Comment (zarirhamza) [inline contrib/labstack/echo.v4/option.go:45]: Also we should document in the comments here that by default we do no translation which is why we're setting the value to nil.

**Review by zarirhamza — CHANGES_REQUESTED** [review]
Few quick naming and documentation changes and after should be good

File: contrib/labstack/echo.v4/echotrace.go [hunk 2]
```diff
@@ -88,8 +88,8 @@ func Middleware(opts ...Option) echo.MiddlewareFunc {
 
 				// It is impossible to determine what the final status code of a request is in echo.
 				// This is the best we can do.
-				var echoErr *echo.HTTPError
-				if errors.As(err, &echoErr) {
+				var echoErr *echo.HTTPError = cfg.customErrorFunc(err)
+				if echoErr != nil || errors.As(err, &echoErr) {
```
Comment (rarguelloF) [inline contrib/labstack/echo.v4/echotrace.go:92]:
```suggestion
				if echoErr, ok := cfg.translateError(err); ok {
```

File: contrib/labstack/echo.v4/echotrace_test.go [hunk 3]
```diff
@@ -424,6 +424,57 @@ func TestIgnoreRequestFunc(t *testing.T) {
...
+type testCustomError struct {
+	TestCode int
+}
+
+// Error satisfies the apierror interface
+func (e *testCustomError) Error() string {
+	return "test"
+}
+
+func TestCustomErrorFunc(t *testing.T) {
```
Comment (rarguelloF) [inline contrib/labstack/echo.v4/echotrace_test.go:436]:
```suggestion
func TestWithErrorTranslator(t *testing.T) {
```

File: contrib/labstack/echo.v4/echotrace_test.go [hunk 4]
```diff
@@ -424,6 +424,57 @@ func TestIgnoreRequestFunc(t *testing.T) {
...
+	defer mt.Stop()
+	var called, traced bool
+
+	// setup
+	customErrorFunc := func(e error) *echo.HTTPError {
+		return &echo.HTTPError{
+			Message: e.(*testCustomError).Error(),
+			Code:    e.(*testCustomError).TestCode,
+		}
+	}
```
Comment (rarguelloF) [inline contrib/labstack/echo.v4/echotrace_test.go:448]:
```suggestion
	translateError := func(e error) (*echo.HTTPError, bool) {
		return &echo.HTTPError{
			Message: e.(*testCustomError).Error(),
			Code:    e.(*testCustomError).TestCode,
		}
	}, true
```

**Review by zarirhamza — APPROVED** [review]

File: contrib/labstack/echo.v4/option.go (hunk 1 above)
Comment (rarguelloF) [inline contrib/labstack/echo.v4/option.go:45]:
```suggestion
	cfg.translateError = func(err error) (*echo.HTTPError, bool) {
               var echoErr *echo.HTTPError
               if errors.As(err, &echoErr) {
                       return echoErr, true
               }
               return nil, false
       }
```

File: contrib/labstack/echo.v4/option.go [hunk 5]
```diff
@@ -33,11 +34,15 @@ type Option func(*config)
 // IgnoreRequestFunc determines if tracing will be skipped for a request.
 type IgnoreRequestFunc func(c echo.Context) bool
 
+// CustomErrorFunc attempts to translate a non-standard echo error to the required format
+type CustomErrorFunc func(err error) *echo.HTTPError
```
Comment (rarguelloF) [inline contrib/labstack/echo.v4/option.go:38]:
```suggestion
```

File: contrib/labstack/echo.v4/option.go [hunk 6]
```diff
@@ -24,6 +24,7 @@ type config struct {
 	noDebugStack      bool
 	ignoreRequestFunc IgnoreRequestFunc
 	isStatusError     func(statusCode int) bool
+	customErrorFunc   CustomErrorFunc
```
Comment (rarguelloF) [inline contrib/labstack/echo.v4/option.go:27]:
```suggestion
	translateError    func(err error) (*echo.HTTPError, bool)
```

File: contrib/labstack/echo.v4/option.go [hunk 7]
```diff
@@ -87,6 +92,14 @@ func WithIgnoreRequest(ignoreRequestFunc IgnoreRequestFunc) Option {
 	}
 }
 
+// WithCustomErrorFunction sets a function to translate custom errors into the
+// expected format
+func WithCustomErrorFunc(customErrorFunc CustomErrorFunc) Option {
```
Comment (rarguelloF) [inline contrib/labstack/echo.v4/option.go:97]:
```suggestion
// WithErrorTranslator sets a function to translate Go errors into echo Errors.
// This is used for extracting the HTTP response status code.
func WithErrorTranslator(fn func(err error) (*echo.HTTPError, bool)) Option {
```

**Review by rarguelloF — COMMENTED** [review]
Thanks for the contribution! @mattscamp

After reviewing with @zarirhamza, we just came up with a few minor changes to simplify the code a little bit and make the name of the option a little bit more descriptive. Once these few changes are made, it should be good to go.

**Comment by mattscamp** [comment]
Thanks @rarguelloF @zarirhamza - made those suggested updates!

**Review by zarirhamza — APPROVED** [review]
LGTM

**Review by rarguelloF — APPROVED** [review]

**Review by ajgajg1134 — APPROVED** [review]
Looks good to me! Thanks!!

**Pull request merged**